### PR TITLE
Pairwise distances: first implementation

### PR DESCRIPTION
**Pull request by mbatoul**

### CODE REVIEW

File: .github/workflows/benchmarks.yml [hunk 1]
```diff
@@ -28,7 +24,7 @@ jobs:
       - name: Setup miniconda
```
Comment (jjerphan) [inline .github/workflows/benchmarks.yml:24]: Small reminder for cache invalidation (defined on the step just above): you might want to invalidate it when your implementation changes.

File: benchmarks.ipynb [hunk 2]
```diff
@@ -2,11 +2,170 @@
  "cells": [
```
Comment (jjerphan) [inline benchmarks.ipynb]: A good way to remove diff from notebook which aren't made to be visualized statically is to clean `outputs` on the UI priorly to adding changes.

File: benchmarks/__main__.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        bench_instance.run()
+
+    env_info = {}
+    env_info["system_info"] = _get_sys_info()
+    env_info["dependencies_info"] = _get_deps_info()
+    env_info["threadpool_info"] = threadpool_info()
+    env_info["cpu_count"] = joblib.cpu_count(only_physical_cores=True)
+
+    with open("benchmarks/env_info.txt", "w") as f:
+        json.dump(env_info, f)
```
Comment (jjerphan) [inline benchmarks/__main__.py:26]: That's indeed something you'd like to keep :+1:

File: benchmarks/core.py [hunk 4]
```diff
@@ -6,12 +6,15 @@
 
 
 class BaseBenchmark:
+    def __init__(self, name):
+        self.name = name
+
```
Comment (jjerphan) [inline benchmarks/core.py]: Small nitpick to prefer private attributes:
```suggestion
    def __init__(self, name):
        self._name = name
```

File: benchmarks/core.py [hunk 5]
```diff
@@ -6,12 +6,15 @@
...
     def _run(self):
         raise NotImplementedError
 
     def _to_csv(self):
         self.results_.to_csv(
-            "benchmarks/results/results.csv",
+            f"benchmarks/results/{self.name}.csv",
```
Comment (jjerphan) [inline benchmarks/core.py]: If the previous suggestion applies.
```suggestion
            f"benchmarks/results/{self._name}.csv",
```

File: benchmarks/pairwise_dist.py [hunk 6]
```diff
@@ -6,7 +6,8 @@
 
 
 class Benchmark(BaseBenchmark):
-    def __init__(self, functions, datasets):
+    def __init__(self, name, functions, datasets):
+        self.name = name
```
Comment (jjerphan) [inline benchmarks/pairwise_dist.py]: For OO design consistency, calling `BaseBenchmark`'s constructor is better:
```suggestion
        super().__init__(name)
```

File: benchmarks/pairwise_dist.py [hunk 7]
```diff
@@ -6,7 +6,8 @@
...
         self.functions = functions
         self.datasets = datasets
```
Comment (jjerphan) [inline benchmarks/pairwise_dist.py]: You can set those privates as well. If you need to have public setters and setters, you can use [properties](https://docs.python.org/3/library/functions.html#property).

File: environment.yml [hunk 8]
```diff
@@ -2,18 +2,16 @@ name: cython_experimentations
...
   - jupyter
   - numpy
+  - scikit-learn
   - cython
   - matplotlib
   - seaborn
   - pandas
   - pyaml
-  - joblib
-  - scipy
```
Comment (jjerphan) [inline environment.yml:14]: Don't you need scipy?

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+        sources=["src/pairwise_dist/_pairwise_dist.pyx"],
+        include_dirs=[numpy.get_include()],
+        extra_compile_args=["-fopenmp", "-O3", "-ftree-vectorize"],
+        extra_link_args=["-fopenmp"],
+        define_macros=[("NPY_NO_DEPRECATED_API", "NPY_1_7_API_VERSION")],
+    )
+]
+
+setup(
+    name="cython_experimentations",
```
Comment (jjerphan) [inline setup.py:17]: You can add yourself as author here! :slightly_smiling_face:

File: src/pairwise_dist/_pairwise_dist.pyx [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+from libc.math cimport fabs
+from cython.parallel cimport prange
+from cython cimport floating, integral
+cimport cython
+
+cdef void _compute_dist(
+    floating[::1] X_a_row,
+    floating[::1] X_b_row,
+    integral n_features,
+    floating *dist
```
Comment (jjerphan) [inline src/pairwise_dist/_pairwise_dist.pyx]: You can always add a trailing comma to the parameters list.
```suggestion
    floating *dist,
```

It can helps with minimising line diffs when changing it.
This applies to all sorts of list (including imports, in-place dictionary definition, etc.)

File: src/pairwise_dist/_pairwise_dist.pyx [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+) nogil:
+    cdef integral i
+    for i in range(n_features):
```
Comment (jjerphan) [inline src/pairwise_dist/_pairwise_dist.pyx]: Small nitpick for readability.
```suggestion
    cdef integral i

    for i in range(n_features):
```

File: src/pairwise_dist/_pairwise_dist.pyx [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        dist[0] += fabs(X_a_row[i] - X_b_row[i])
+
+cdef void _pairwise_dist(
+    floating[:, ::1] X_a, # IN
+    floating[:, ::1] X_b, # IN
+    floating[:, ::1] distances, # OUT
+    integral dummy
```
Comment (jjerphan) [inline src/pairwise_dist/_pairwise_dist.pyx]: As discussed, I don't know what's the best here to make Cython able to specialize functions for integral.
Using `int` is or even `cnp.intp_t` types should be fine as we solely used them for indexing.

File: src/pairwise_dist/_pairwise_dist.pyx [hunk 13]
```diff
@@ -0,0 +1,46 @@
+# cython: boundscheck=False
```
Comment (jjerphan) [inline src/pairwise_dist/_pairwise_dist.pyx:3]: There are others checks that you can remove for efficiency but I am sure that you know this already. :)

**Review by jjerphan — COMMENTED** [review]
That's a nice first start! :clap: 
---
The generated `.c` file is still present. You might want to erase it from your local repo cache using something like:
```sh
git rm --cached **/*.c
```
---

I left some comment, but in overall it's good: the Python API is concise and the implementation is done in `cdef` functions which should map to C code with no much interaction with the CPython interpreter.

Looking forward to your future iterations. :)

File: src/pairwise_dist/_pairwise_dist.pyx [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+cimport numpy as np
+import numpy as np
+from libc.math cimport fabs
+from cython.parallel cimport prange
+from cython cimport floating, integral
+cimport cython
+
+cdef void _compute_dist(
```
Comment (jeremiedbb) [inline src/pairwise_dist/_pairwise_dist.pyx]: It's L1 dist here, maybe rename this function to make it clear

File: src/pairwise_dist/_pairwise_dist.pyx (hunk 12 above)
Comment (jeremiedbb) [reply src/pairwise_dist/_pairwise_dist.pyx]: Yes I think using int is perfectly fine here for now. Adding a unused param feels sketchy and is a good indicator that this should not be fused typed

File: src/pairwise_dist/_pairwise_dist.pyx [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+) nogil:
+    cdef:
+        integral i, j
+        integral n_rows_X_a = X_a.shape[0]
+        integral n_rows_X_b = X_b.shape[0]
+        integral n_features = X_a.shape[1]
+        
+    for i in prange(n_rows_X_a, nogil=True):
+        for j in range(n_rows_X_b):
+            _compute_dist(X_a[i], X_b[j], n_features, &distances[i, j])
```
Comment (jeremiedbb) [inline src/pairwise_dist/_pairwise_dist.pyx]: in this implementation passing a pointer for distances[i, j] seems unecessary. You could just do
``distances[i, j] = _compute(dist(X_a[i], X_b[j], n_features))``

File: src/pairwise_dist/_pairwise_dist.pyx [hunk 16]
```diff
@@ -0,0 +1,47 @@
+# cython: boundscheck=False
+
+cimport numpy as np
+import numpy as np
+from libc.math cimport fabs
```
Comment (jeremiedbb) [inline src/pairwise_dist/_pairwise_dist.pyx:12]: did you try using the builtin `abs` from python ? I think cython falls back to the C implementation of abs in that case

**Review by jeremiedbb — COMMENTED** [review]
The link for the benchmarks returns a 404.

The implementation looks good. You can now compare with ``pairwise_distances(metric="l1")`` from sklearn on datasets with different sizes (on single thread and then on multi-thread).

File: src/pairwise_dist/_pairwise_dist.pyx (hunk 15 above)
Comment (jeremiedbb) [reply src/pairwise_dist/_pairwise_dist.pyx]: See this issue https://github.com/scikit-learn/scikit-learn/issues/17299. It looks like a good reproducer for it. Is it faster to do as explained in the issue ?

File: src/pairwise_dist/_pairwise_dist.pyx (hunk 16 above)
Comment (jjerphan) [reply src/pairwise_dist/_pairwise_dist.pyx:12]: IIRC, Cython failed because `abs` take `integrals` as inputs.

Comment (jeremiedbb) [reply src/pairwise_dist/_pairwise_dist.pyx:12]: I don't understand. This works
```
%%cython
cdef double a = -1.234
cdef double b = abs(a)
print(b)
```

Comment (jeremiedbb) [reply src/pairwise_dist/_pairwise_dist.pyx:12]: I'm not talking about abs from C standard library, but abs from python builtins

Comment (jjerphan) [reply src/pairwise_dist/_pairwise_dist.pyx:12]: Can python's abs be used within `nogil` contexts?

Comment (jeremiedbb) [reply src/pairwise_dist/_pairwise_dist.pyx:12]: hmm it can't so you can forget about this comment

File: src/pairwise_dist/_pairwise_dist.pyx (hunk 15 above)
Comment (mbatoul) [reply src/pairwise_dist/_pairwise_dist.pyx]: It seems faster indeed! Some quick benchmarks:

image

We will see how it improves performance compared to `scipy` and `scikit-learn`.

File: src/pairwise_dist/_pairwise_dist.pyx [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+) nogil:
+    cdef:
+        int i, j
+        int n_rows_X_a = X_a.shape[0]
+        int n_rows_X_b = X_b.shape[0]
+        int n_features = X_a.shape[1]
+        
+    for i in prange(n_rows_X_a, nogil=True):
+        for j in range(n_rows_X_b):
+            distances[i, j] = _l1_distance(X_a, i, X_b, j, n_features)
```
Comment (jjerphan) [inline src/pairwise_dist/_pairwise_dist.pyx:46]: You can inspect in another PR the parallelization strategy when `n_rows_X_b >> n_rows_X_a`.

**Review by jjerphan — APPROVED** [review]
We are good with this PR!

**Review by quanghuynh10111-png — COMMENTED** [review]
[](url)
